cpp/src/arrow/csv/reader_test.cc

```diff
@@ -0,0 +1,169 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <cstdint>
+#include <string>
+#include <thread>
+#include <utility>
+#include <vector>
+
+#include <gtest/gtest.h>
+
+#include "arrow/csv/options.h"
+#include "arrow/csv/reader.h"
+#include "arrow/csv/test_common.h"
+#include "arrow/io/interfaces.h"
+#include "arrow/status.h"
+#include "arrow/table.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/util/future.h"
+#include "arrow/util/thread_pool.h"
+
+namespace arrow {
+namespace csv {
+
+void StressTableReader(
+    std::function<Result<std::shared_ptr<TableReader>>(std::shared_ptr<io::InputStream>)>
+        reader_factory) {
+  const int NTASKS = 100;
+  const int NROWS = 1000;
+  ASSERT_OK_AND_ASSIGN(auto table_buffer, MakeSampleCsvBuffer(NROWS));
+
+  std::vector<Future<std::shared_ptr<Table>>> task_futures(NTASKS);
+  for (int i = 0; i < NTASKS; i++) {
+    auto input = std::make_shared<io::BufferReader>(table_buffer);
+    ASSERT_OK_AND_ASSIGN(auto reader, reader_factory(input));
+    task_futures[i] = reader->ReadAsync();
+  }
+  auto combined_future = All(task_futures);
+  combined_future.Wait();
+
+  ASSERT_OK_AND_ASSIGN(std::vector<Result<std::shared_ptr<Table>>> results,
+                       combined_future.result());
+  for (auto&& result : results) {
+    ASSERT_OK_AND_ASSIGN(auto table, result);
+    ASSERT_EQ(NROWS, table->num_rows());
+  }
+}
+
+void TestNestedParallelism(
+    std::shared_ptr<internal::ThreadPool> thread_pool,
+    std::function<Result<std::shared_ptr<TableReader>>(std::shared_ptr<io::InputStream>)>
+        reader_factory) {
+  const int NROWS = 1000;
+  ASSERT_OK_AND_ASSIGN(auto table_buffer, MakeSampleCsvBuffer(NROWS));
+  auto input = std::make_shared<io::BufferReader>(table_buffer);
+  ASSERT_OK_AND_ASSIGN(auto reader, reader_factory(input));
+
+  Future<std::shared_ptr<Table>> table_future;
+
+  auto read_task = [&reader, &table_future]() mutable {
+    table_future = reader->ReadAsync();
+    return Status::OK();
+  };
+  ASSERT_OK_AND_ASSIGN(auto future, thread_pool->Submit(read_task));
+  ASSERT_TRUE(future.Wait(1));
+
+  if (future.is_finished()) {
+    ASSERT_TRUE(table_future.Wait(1));
+    if (table_future.is_finished()) {
+      ASSERT_OK_AND_ASSIGN(auto table, table_future.result());
+      ASSERT_EQ(table->num_rows(), NROWS);
+    }
+  }
+}  // namespace csv
+
+TEST(SerialReaderTests, Stress) {
+  auto task_factory = [](std::shared_ptr<io::InputStream> input_stream) {
+    return TableReader::Make(default_memory_pool(), io::AsyncContext(), input_stream,
+                             ReadOptions::Defaults(), ParseOptions::Defaults(),
+                             ConvertOptions::Defaults());
+  };
+  StressTableReader(task_factory);
+}
+
+TEST(SerialReaderTests, NestedParallelism) {
+  ASSERT_OK_AND_ASSIGN(auto thread_pool, internal::ThreadPool::Make(1));
```

Comment:
Well, this is the serial reader, so it doesn't make use of a thread pool.  However, I could pass one in the `io::AsyncContext` anyway for good measure.  It's not harmless in the async or the threaded (commented out) cases because the thread pool is passed in via the `io::AsyncContext` and is reused by the readers.  There is a reason the threaded case is commented out (it deadlocks).